Metrics processor

HudsonHumphries opened the pull request:
This PR is the processor for the Metrics Pipeline. This processor can either be stateless or stateful. If the processor is stateless, it sends metric deltas to the exporter. If the processor is stateful, it sends metric cumulatives to the exporter. When the processor keeps state, it will create a copy of the aggregator passed in and will continue to update this copied aggregator every time the same one is passed in. The key to find this copied aggregator is a string of the data inside the record that holds the aggregator. This is a temporary solution that should be replaced by a class with a custom hash function. This processor should be created by the user and set in the constructor of the controller.

**This PR contains code that is yet to be merged into the repository and should not be reviewed. Please only review the processor portion of this code. Thank you**

HudsonHumphries commented:
@reyang @pyohannes This PR should be ready for review and can be merged once the Sketch Aggregator is merged.

reyang commented on sdk/include/opentelemetry/sdk/metrics/ungrouped_processor.h:
```diff
@@ -0,0 +1,238 @@
+    auto ins_kind = aggregator->get_instrument_kind();
+    auto agg_kind = aggregator->get_aggregator_kind();
+
+    if (agg_kind == sdkmetrics::AggregatorKind::Counter)
```
Minor: wonder if a switch-case statement would make it easier to read?

reyang commented on sdk/include/opentelemetry/sdk/metrics/ungrouped_processor.h:
```diff
@@ -0,0 +1,238 @@
+}  // namespace metrics
+}  // namespace sdk
+
+OPENTELEMETRY_END_NAMESPACE
```
nit: add an empty line.

reyang commented on sdk/include/opentelemetry/sdk/metrics/ungrouped_processor.h:
```diff
@@ -0,0 +1,238 @@
+  }
+
+  std::map<metrics_api::InstrumentKind, std::string> ins_to_string{
+      {metrics_api::InstrumentKind::Counter, "Counter"},
```
minor: consider using macro to simplify this (string tagging https://docs.microsoft.com/cpp/preprocessor/stringizing-operator-hash).

HudsonHumphries replied:
I've never seen something like this, its a much better solution. Thanks!

reyang commented on sdk/src/metrics/ungrouped_processor.cc:
```diff
@@ -0,0 +1,173 @@
+    std::string key               = iter.first;
+    std::size_t description_index = key.find("/description/");
+    std::size_t labels_index      = key.find("/labels/");
+    std::size_t instrument_index  = key.find("/instrument/");
```
Probably leave to TODO here for micro-optimization (instead of searching 3 strings in 3 scans, we can potentially scan once using a [DFA](https://en.wikipedia.org/wiki/Deterministic_finite_automaton)).

HudsonHumphries replied:
Yeah this could definitely be done better, I was thinking of adding a future enhancement of using a struct to hold these values with a custom hash function instead of using this string approach. Thoughts on this?

HudsonHumphries commented:
Updated based off @reyang feedback, rebased with master. Bazel asan is failing from the SketchAggregator. Once this is fixed and merged in, this PR should be ready for merge if there are no more comments.

reyang commented on sdk/include/opentelemetry/sdk/metrics/aggregator/sketch_aggregator.h:
```diff
@@ -0,0 +1,255 @@
+
+}  // namespace metrics
+}  // namespace sdk
+OPENTELEMETRY_END_NAMESPACE
```
nit: would you add a blank line before EOF for all the files in this PR?

HudsonHumphries replied:
fixed

reyang commented on sdk/include/opentelemetry/sdk/metrics/ungrouped_processor.h:
```diff
@@ -0,0 +1,246 @@
+#include "opentelemetry/sdk/metrics/record.h"
+#include "opentelemetry/version.h"
+
+#define stringer(x) (#x)
```
It'll cause macro pollution.

HudsonHumphries replied:
I've never used macros before, how should I change this?

HudsonHumphries replied:
Do I just need to #undef string at EOF?

reyang replied:
In addition, avoid a common name as it might cause conflict with other headers.

e.g. `#define UNGROUPED_PROCESSOR_STRINGER`.

Stepping back, need to evaluate the pros and cons and if it doesn't pay much, don't use it 😄

HudsonHumphries replied:
Moved it over to ungrouped_processor.cc, let me know if this is better. I think it pays a little bit as we don't have to create a map.

HudsonHumphries commented:
This PR should be ready to merge if I addressed all of your comments @reyang
